### PR TITLE
fix: clear expired events workflow execution faild

### PR DESCRIPTION
**Pull request by amliuyong**
close https://github.com/awslabs/clickstream-analytics-on-aws/issues/289

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

## Summary

- fix https://github.com/awslabs/clickstream-analytics-on-aws/issues/289
- fix upgrade error for alarm 'Clickstream|test_project_007 Max file age d9170ff0 already exists'
- fix upgrade error for s3 prefix of ods events files from control plane

(describe what this merge request does)

## Implementation highlights

(describe how the merge request does for feature changes, share the RFC link if it has)

## Test checklist

- [x] add new test cases
- [x] all code changes are covered by unit tests
- [ ] end-to-end tests
  - [ ] deploy web console with CloudFront + S3 + API gateway
  - [ ] deploy web console within VPC
  - [ ] deploy ingestion server
    - [ ] with MSK sink
    - [ ] with KDS sink
    - [ ] with S3 sink
  - [x] deploy data processing
  - [x] deploy data modeling
    - [x] new Redshift Serverless
    - [ ] provisioned Redshift
    - [ ] Athena
  - [ ] deploy with reporting

## Is it a breaking change

- [ ] add parameters without default value in stack
- [ ] introduce new service permission in stack
- [ ] introduce new top level stack module

## Miscellaneous

- [ ] introduce new symbol link source file(s) to be shared among infra code, web console frontend, and web console backend

### CODE REVIEW

File: src/analytics/lambdas/clear-expired-events-workflow/check-clear-status.ts [hunk 1]
```diff
@@ -53,13 +53,12 @@ export const handler = async (event: ClearExpiredEventsEvent) => {
     };
   } else if (response.Status == StatusString.FAILED || response.Status == StatusString.ABORTED) {
     logger.info(`Executing ${queryId} status of statement is ${response.Status}`);
-    const queryResult = await queryClearLog(appId);
```
Comment (zxkane) [inline src/analytics/lambdas/clear-expired-events-workflow/check-clear-status.ts:56]: Is the sp log unavailable?

File: src/analytics/lambdas/custom-resource/get-source-prefix.ts [hunk 2]
```diff
@@ -47,14 +48,18 @@ async function _handler(event: CdkCustomResourceEvent) {
   const odsEventPrefix = (event.ResourceProperties as PropertiesType).odsEventPrefix;
 
   let prefix = odsEventPrefix;
+  const projectId = (event.ResourceProperties as PropertiesType).projectId;
 
   if (odsEventPrefix.endsWith('/ods_events/') || odsEventPrefix.endsWith('/ods_events')) {
     prefix = odsEventPrefix.replace(new RegExp('/ods_events[/]?'), '');
+  } else if (odsEventPrefix.endsWith(`clickstream/${projectId}/data/ods/`)) {
+    // control plane set the prefix to clickstream/<projectId>/data/ods/
+    // the actual prefix should be clickstream/<projectId>/data/ods/<projectId>/
```
Comment (zxkane) [inline src/analytics/lambdas/custom-resource/get-source-prefix.ts:57]: Does the spark ETL add the additional prefix `projectId`?

File: src/analytics/private/sqls/redshift/sp-scan-metadata.sql [hunk 3]
```diff
@@ -1,5 +1,6 @@
-CREATE OR REPLACE PROCEDURE {{schema}}.sp_scan_metadata(top_frequent_properties_limit NUMERIC, day_range NUMERIC)
-AS $$ 
+CREATE OR REPLACE PROCEDURE {{schema}}.sp_scan_metadata(top_frequent_properties_limit NUMERIC, day_range NUMERIC) 
+NONATOMIC AS
```
Comment (zxkane) [inline src/analytics/private/sqls/redshift/sp-scan-metadata.sql:2]: For this I'm not sure, if we can upsert the existing items when re-run the procedure?

@dengmingtong could you confirm if this change it's safe?

File: src/analytics/private/sqls/redshift/sp-upsert-users.sql [hunk 4]
```diff
@@ -1,5 +1,5 @@
-CREATE OR REPLACE PROCEDURE {{schema}}.{{sp_upsert_users}}()
-AS 
+CREATE OR REPLACE PROCEDURE {{schema}}.{{sp_upsert_users}}() 
+NONATOMIC AS
```
Comment (zxkane) [inline src/analytics/private/sqls/redshift/sp-upsert-users.sql:2]: please make sure it's safe for re-running the procedure again.

**Review by zxkane — CHANGES_REQUESTED** [review]

File: src/analytics/lambdas/clear-expired-events-workflow/check-clear-status.ts (hunk 1 above)
Comment (amliuyong) [reply src/analytics/lambdas/clear-expired-events-workflow/check-clear-status.ts:56]: want to keep log when Status == FAILED/ABORTED, queryClearLog will clean the error log

File: src/analytics/private/sqls/redshift/sp-scan-metadata.sql (hunk 3 above)
Comment (amliuyong) [reply src/analytics/private/sqls/redshift/sp-scan-metadata.sql:2]: confirmed by @dengmingtong

File: src/analytics/private/sqls/redshift/sp-upsert-users.sql (hunk 4 above)
Comment (amliuyong) [reply src/analytics/private/sqls/redshift/sp-upsert-users.sql:2]: yes.

File: src/analytics/lambdas/custom-resource/get-source-prefix.ts (hunk 2 above)
Comment (amliuyong) [reply src/analytics/lambdas/custom-resource/get-source-prefix.ts:57]: ETL saved the files to <prefix>/<projectId>/ods_events/

I supposed that control plane set the prefix ` <prefix>/<projectId>/ods_events/`, but actually, it set the prefix to `<prefix>/`, this bug was reported by @dengmingtong  this morning. 

the <prefix> is 'clickstream/<projectId>/data/ods/'

**Review by zxkane — APPROVED** [review]

**Review by zxkane — APPROVED** [review]

**Pull request merged**